n3fit/src/n3fit/layers/DY.py

```diff
@@ -19,48 +32,35 @@ def gen_mask(self, basis):
+            fk = op.einsum('nFxy -> nFyx', fk)
+            mask_and_fk = (mask, fk)
+            return mask_and_fk
 
-        First uses the basis of active combinations to generate a luminosity tensor
-        with only some flavours active.
+    def build(self, input_shape):
+        super().build(input_shape)
+        if self.num_replicas > 1:
 
-        The concatenate function returns a rank-3 tensor (combination_index, xgrid, xgrid)
-        which can in turn be contracted with the rank-4 fktable.
+            def compute_observable(pdf, masked_fk):
+                temp = op.einsum('nxfyg, bryg -> brnxf', masked_fk, pdf)
+                return op.einsum('brnxf, brxf -> brn', temp, pdf)
 
-        Parameters
-        ----------
-            pdf_in: tensor
-                rank 4 tensor (batchsize, replicas, xgrid, flavours)
+        else:
 
-        Returns
-        -------
-            results: tensor
-                rank 3 tensor (batchsize, replicas, ndata)
-        """
-        # Hadronic observables might need splitting of the input pdf in the x dimension
-        # so we have 3 different paths for this layer
+            def compute_observable(pdf, mask_and_fk):
+                # with 1 replica, it's more efficient to mask the PDF rather than the fk table
+                mask, unmasked_fk = mask_and_fk
+                pdf = pdf[0][0]  # yg
 
-        results = []
-        if self.many_masks:
-            if self.splitting:
-                splitted_pdf = op.split(pdf_raw, self.splitting, axis=2)
-                for mask, pdf, fk in zip(self.all_masks, splitted_pdf, self.fktables):
-                    pdf_x_pdf = op.pdf_masked_convolution(pdf, mask)
-                    res = op.tensor_product(fk, pdf_x_pdf, axes=[(1, 2, 3), (1, 2, 3)])
-                    results.append(res)
-            else:
-                for mask, fk in zip(self.all_masks, self.fktables):
-                    pdf_x_pdf = op.pdf_masked_convolution(pdf_raw, mask)
-                    res = op.tensor_product(fk, pdf_x_pdf, axes=[(1, 2, 3), (1, 2, 3)])
-                    results.append(res)
-        else:
-            pdf_x_pdf = op.pdf_masked_convolution(pdf_raw, self.all_masks[0])
-            for fk in self.fktables:
-                res = op.tensor_product(fk, pdf_x_pdf, axes=[(1, 2, 3), (1, 2, 3)])
-                results.append(res)
+                mask_x_pdf = op.tensor_product(mask, pdf, axes=[(2,), (1,)])  # Ffg, yg -> Ffy
+                pdf_x_pdf = op.tensor_product(mask_x_pdf, pdf, axes=[(1,), (1,)])  # Ffy, xf -> Fyx
+                # nFyx, Fyx -> n
+                observable = op.tensor_product(unmasked_fk, pdf_x_pdf, axes=[(1, 2, 3), (0, 1, 2)])
+
+                return op.batchit(op.batchit(observable))  # brn
 
-        # the masked convolution removes the batch dimension
-        ret = op.transpose(self.operation(results))
-        return op.batchit(ret)
+        self.compute_observable = compute_observable
```

Comment:
same comment as before, having these function as module level functions would be great

I would also expect a speed up and memory reduction since they will be shared by different observables... You might want to put a `@tf.function` decorator on them.

... unless the speed up is coming from a memory-tradeoff by having these functions be observable specific? But I would hope not...
if that were the case you can still compile them when you attach them to the given layer by doing

`self.compute_observable = @tf.function(compute_observable)`